src/lib/server/server.cpp

```diff
@@ -1,40 +1,43 @@
 #include "server.hpp"
 
-#include <boost/asio/placeholders.hpp>
-#include <boost/bind.hpp>
-
-#include "client_connection.hpp"
-#include "server_session.hpp"
-#include "task_runner.hpp"
-#include "then_operator.hpp"
+#include <iostream>
+#include <thread>
 
 namespace opossum {
 
-using opossum::then_operator::then;
-
-Server::Server(boost::asio::io_service& io_service, uint16_t port)
-    : _io_service(io_service),
-      _acceptor(io_service, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), port)),
-      _socket(io_service) {
-  _accept_next_connection();
+// Specified port (default: 5432) will be opened after initializing the _acceptor
+Server::Server(const uint16_t port, const bool send_execution_info)
+    : _acceptor(_io_service, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), port)),
+      _send_execution_info(send_execution_info) {
+  std::cout << "Server starting on port " << get_port() << std::endl;
 }
 
-void Server::_accept_next_connection() {
-  _acceptor.async_accept(_socket, boost::bind(&Server::_start_session, this, boost::asio::placeholders::error));
+void Server::_accept_new_session() {
+  // Create a new session. This will also open a new data socket in order to communicate with the client
+  // For more information on TCP ports + Asio see:
+  // https://www.gamedev.net/forums/topic/586557-boostasio-allowing-multiple-connections-to-a-single-server-socket/
+  auto new_session = std::make_shared<Session>(_io_service, _send_execution_info);
+  _acceptor.async_accept(*(new_session->get_socket()),
+                         boost::bind(&Server::_start_session, this, new_session, boost::asio::placeholders::error));
 }
 
-void Server::_start_session(boost::system::error_code error) {
+void Server::_start_session(const std::shared_ptr<Session>& new_session, const boost::system::error_code& error) {
   if (!error) {
-    auto connection = std::make_shared<ClientConnection>(std::move(_socket));
-    auto task_runner = std::make_shared<TaskRunner>(_io_service);
-    auto session = std::make_shared<ServerSession>(connection, task_runner);
-    // Start the session and release it once it has terminated
-    session->start() >> then >> [=]() mutable { session.reset(); };
+    std::thread session_thread([new_session] { new_session->start(); });
```

Comment:
I'm admittedly quite late to the party, but spawning an undefined number of threads outside of the scheduler does not seem right.

Reply:
@mrks Please note that the current implementation spawns an unbounded amount of threads as well. This new implementation does it in a (way) more sane way. We can talk about the unification in the scheduler and whether this needs to happen as part of this change.